kratos/mpi/utilities/parallel_fill_communicator.cpp

```diff
@@ -0,0 +1,466 @@
+
+    // Fill nodes for LocalMesh and GhostMesh.
+    for (auto it_node = rModelPart.NodesBegin(); it_node != rModelPart.NodesEnd(); ++it_node)
+    {
+        const unsigned index = it_node->FastGetSolutionStepValue(PARTITION_INDEX);
+        if (index == MyRank)
+            r_local_nodes.push_back(*(it_node.base()));
+        else
+            r_ghost_nodes.push_back(*(it_node.base()));
+    }
+
+    // Fill nodes for the InterfaceMesh.
+    for (ModelPart::MeshType& r_color_interface_mesh : rModelPart.GetCommunicator().InterfaceMeshes())
+    {
+        ModelPart::NodesContainerType& r_color_interface_nodes =
+            r_color_interface_mesh.Nodes();
+        for (auto it = r_color_interface_nodes.begin(); it != r_color_interface_nodes.end(); ++it)
+            r_interface_nodes.push_back(*(it.base()));
+    }
+    r_interface_nodes.Unique();
+    r_local_nodes.Unique();
+    r_ghost_nodes.Unique();
+
+    // Assign elements and conditions for LocalMesh.
+    rModelPart.GetCommunicator().LocalMesh().Elements().clear();
+    rModelPart.GetCommunicator().LocalMesh().Conditions().clear();
+    rModelPart.GetCommunicator().LocalMesh().Elements() = rModelPart.Elements();
+    rModelPart.GetCommunicator().LocalMesh().Conditions() = rModelPart.Conditions();
+
+    // Call the sub model part.
+    for (ModelPart& r_sub_model_part : rModelPart.SubModelParts())
+        ComputeCommunicationPlan(r_sub_model_part);
+
+    KRATOS_CATCH("");
+}
+
+void ParallelFillCommunicator::GenerateMeshes(int NeighbourPID, int MyPID, unsigned Color, ModelPart& rModelPart)
+{
+    KRATOS_TRY;
+
+    KRATOS_ERROR_IF(NeighbourPID == MyPID)
+        << "Neighbour process coincides with rank! this should not happen."
+        << std::endl;
+
+    if (NeighbourPID == -1) // Don't communicate with this process.
+        return;
+
+    const DataCommunicator& r_data_communicator = rModelPart.GetCommunicator().GetDataCommunicator();
+
+    ModelPart::NodesContainerType& r_ghost_nodes =
+        rModelPart.GetCommunicator().GhostMesh(Color).Nodes();
+    r_ghost_nodes.clear();
+
+    // Fill nodes for GhostMesh(Color).
+    for (auto it = rModelPart.NodesBegin(); it != rModelPart.NodesEnd(); ++it)
+    {
+        const int index = it->FastGetSolutionStepValue(PARTITION_INDEX);
+        if (index == NeighbourPID)
+            r_ghost_nodes.push_back(*(it.base()));
+    }
+    unsigned num_ghost_nodes = r_ghost_nodes.size();
+    r_ghost_nodes.Unique();
+    KRATOS_ERROR_IF(num_ghost_nodes != r_ghost_nodes.size())
+        << "The list of nodes to receive has repeated nodes." << std::endl;
+
+    std::vector<int> ids_to_receive(r_ghost_nodes.size());
+
+    { // Fill receive ids (ids of ghost nodes).
+        int i = 0;
+        for (const ModelPart::NodeType& rNode : r_ghost_nodes)
+            ids_to_receive[i++] = rNode.Id();
+    }
+
+    std::vector<int> ids_to_send;
+    { // Syncronize how many nodes need to be sent/received.
```

Comment:
why protected?

Reply:
To ensure that the tags are deallocated once they go out of scope and are no longer needed. By the way, this was already like this, I kept it because it is correct code.